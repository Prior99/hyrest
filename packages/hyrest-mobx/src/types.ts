export interface Indexable<KeyType> {
    id?: KeyType;
}

export interface Target {
    value: any;
    checked?: boolean;
    type: string;
}

export interface ReactEvent {
    target: Target;
}

/**
 * A context factory, being provided to `@hasFields` to retrieve the context inside
 * the `Field` instances.
 */
export type ContextFactory<TContext> = () => TContext;
