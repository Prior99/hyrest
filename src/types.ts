export interface Answer<T> {
    readonly message?: string;
    readonly data: T;
}

export interface Params {
    [key: string]: string;
}

/**
 * This type represents all possible HTTP methods.
 */
export type HTTPMethod =
    "GET" |
    "POST" |
    "PATCH" |
    "PUT" |
    "DELETE" |
    "HEAD" |
    "OPTIONS" |
    "TRACE";

/**
 * This error will be thrown if the call to a route from the client did not succeed.
 * It transports the HTTP status code as well as the answer from the server if available.
 * Both can be `undefined` if a network failure is the reason for the error.
 */
export class ApiError extends Error {
    /**
     * The HTTP status code. Can be `undefined`.
     */
    public statusCode: number;
    /**
     * The answer from the server. Can be `undefined`.
     */
    public answer: Answer<any>;

    constructor(statusCode: number, answer: Answer<any>) {
        super()/* istanbul ignore next */;
        Error.captureStackTrace(this, ApiError);
        this.statusCode = statusCode;
        this.answer = answer;
    }
}

export type Constructable<T> = {
    new(): T;
};
