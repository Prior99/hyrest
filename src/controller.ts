import { Constructable } from "./types";
import "isomorphic-fetch";
import "reflect-metadata";

import { Route } from "./route";
import { Params, ApiError, HTTPMethod } from "./types";
import { compile } from "path-to-regexp";
import { isBrowser } from "./is-browser";
import { populate } from "./scope";

export enum ControllerMode {
    SERVER = "server",
    CLIENT = "client",
}

export function getDefaultControllerMode() {
    return isBrowser() ? ControllerMode.CLIENT : ControllerMode.SERVER;
}

function buildQueryString(query: Params) {
    const queryString = Object.keys(query).reduce((result, key) => {
        const unencoded = query[key];
        if (typeof unencoded === "undefined" || unencoded === null) {
            return result;
        }
        const value = encodeURIComponent(unencoded);
        return `${result}&${key}=${value}`;
    }, "");
    // Remove first `&`.
    return queryString.substr(1, queryString.length);
}

/**
 * Assembles a full Url from an express url template, url parameters, as base Url and a query parameters.
 *
 * @param urlParameters The parameters fed to the express url template specified in `subUrl`.
 * @param query The query parameters for the url.
 * @param baseUrl The url to which the template url in `subUrl` is relative to.
 * @param subUrl An url template as used in express.
 *
 * @return The assembled Url.
 */
export function buildUrl(urlParameters: Params, query: Params, baseUrl: string, subUrl: string) {
    const queryString = buildQueryString(query);
    // The url parameters need to be explicitly converted to string as they will be transported via
    // the url string.
    const stringifiedParameters = Object.keys(urlParameters).reduce((result, key) => {
        result[key] = `${urlParameters[key]}`;
        return result;
    }, {} as any);

    // Compile the route's pattern (This is the reverse of what express does with url patterns) and
    // feed it the url parameters which had been converted to string.
    const routeString = compile(subUrl)(stringifiedParameters);

    // Assemble the full url.
    return `${baseUrl}${routeString}?${queryString}`;
}

/**
 * A handler which will be called whenever an Api error is encountered.
 *
 * @param error The Api error that was encountered.
 */
export type ErrorHandler = (error: ApiError) => void;

/**
 * Options with which a controller can be configured.
 * The options are used for all instances of a Controller class.
 */
export interface ControllerOptions {
    /**
     * If set to `true` the controller will `throw` whenever an error is encountered.
     * Default is `true`.
     */
    readonly throwOnError?: boolean;
    /**
     * When supplied, this handler will be called for all errors encountered by this controller.
     */
    readonly errorHandler?: ErrorHandler;
    /**
     * A base Url on which the backend can be publicly reached.
     */
    readonly baseUrl?: string;
    /**
     * Can be `ControllerMode.CLIENT` or `ControllerMode.SERVER`. This is autodetected by default
     * and `ControllerMode.CLIENT` will be used in the browser. This determines if the controller
     * will be used to serve the Api or consume it.
     */
    readonly mode?: ControllerMode;
}

export class Controller {
    public throwOnError = true;
    public errorHandler: ErrorHandler;
    public baseUrl: string;
    public mode: ControllerMode = getDefaultControllerMode();

    constructor(options: ControllerOptions) {
        if (options) {
            this.configure(options);
        }
    }

    /**
     * Will be called by `configureController` and applies the given options to this controller.
     */
    public configure(options: ControllerOptions) {
        const { throwOnError, errorHandler, baseUrl, mode } = options;
        if (typeof throwOnError !== "undefined") { this.throwOnError = throwOnError; }
        if (typeof errorHandler !== "undefined") { this.errorHandler = errorHandler; }
        if (typeof baseUrl !== "undefined") { this.baseUrl = baseUrl; }
        if (typeof mode !== "undefined") { this.mode = mode; }
    }

    /**
     * Will be called instead of each route's body when running in `CLIENT` mode. This function
     * will perform the `fetch` against the backend endpoint.
     *
     * @param route The Route meta information for the method for which the functions was called.
     * @param parameters The Url parameters.
     * @param body The body.
     * @param query The query parameters.
     *
     * @return The unwrapped answer from the backend.
     */
    public async wrappedFetch<T>(route: Route, urlParameters: Params, body: any, query: Params): Promise<T> {
        try {
            // Create the url from the given inputs.
            const url = buildUrl(urlParameters, query, this.baseUrl, route.url);

            // Prepare the headers.
            const headers = new Headers();
            headers.append("content-type", "application/json");

            // Perform the actual fetch.
            const response = await fetch(url, {
                body: JSON.stringify(body),
                headers,
                method: route.method,
            });

            // Parse the response from the Api as JSON.
            const answer = await response.json();

            // If the response's status code was a 2xx status code, return the response as it succeeded.
            if (response.ok) {
                const routeData = answer.data;
                if (typeof route.returnType !== "undefined" && typeof route.scope !== "undefined") {
                    if (Array.isArray(routeData)) {
                        return populate(route.scope, Array, route.returnType, routeData);
                    }
                    return populate(route.scope, route.returnType, routeData);
                }
                return routeData;
            }

            // Otherwise, create an error and handle it according to the controller's configuration.
            // This happens by throwing it, which will raise it to the `catch` block below.
            // That block will then perform the actual error handling.
            const error = new ApiError(response.status, answer);
            throw error;
        } catch (error) {
            // Error's might occur as the network might be interrupted or the response might be malformed.
            // Also an error might be thrown by the `try` block above, if a non-2xx status code has been received.
            if (this.errorHandler) {
                this.errorHandler(error);
            }
            if (this.throwOnError) {
                throw error;
            }
        }
    }
}

export function controller(options?: ControllerOptions): ClassDecorator;
export function controller<U, T extends Constructable<U>>(target: T): T;
/**
 * Decorate an individual class as an Api controller.
 *
 * @param options Global options with which this controller can be configured. `configureController` can also set
 *                set and override these options.
 *
 * @return The decorated class.
 */
export function controller<T extends Function>(arg1: ControllerOptions | T): ClassDecorator | T {
    let decorator: (target: T) => T;
    const options = typeof arg1 === "object" ? arg1 : undefined;
    decorator = function(target: T): T {
        Reflect.defineMetadata("api:controller", new Controller(options), target);
        return target;
    };
    if (typeof arg1 === "function") {
        return decorator(arg1 as T);
    }
    return decorator as ClassDecorator;
}
