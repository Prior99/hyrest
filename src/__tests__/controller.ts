import "isomorphic-fetch";

import { Controller } from "../controller";
import { ApiError, HTTPMethod  } from "../types";

let controller: Controller;

beforeEach(() => {
    controller = new Controller({
        baseUrl: "http://example.com",
    });
});

const urlParameters = {
    id: "some-id",
    other: "7",
};
const query = {
    search: "some-search&? -\\/",
    page: "7",
    size: "100",
};
const body = {
    someKey: "some value",
    someArray: [1, 2, 3, 4],
};
const route = {
    target: controller,
    property: "thisMethodDoesNotExist",
    url: "/user/:id/other/:other",
    method: "GET" as HTTPMethod,
};

test("`wrappedFetch` with a successfull response", async () => {
    const mock = jest.fn();
    global.fetch = mock;
    mock.mockReturnValue({
        json: () => ({
            message: "Everything went well.",
            data: {
                someThing: "Something else.",
            },
        }),
        ok: true,
    });

    const result = await controller.wrappedFetch(route, urlParameters, body, query);

    const headers = new Headers();
    headers.append("content-type", "application/json");
    const url = "http://example.com/user/some-id/other/7?search=some-search%26%3F%20-%5C%2F&page=7&size=100";
    expect(mock).toHaveBeenCalledWith(url, {
        body: JSON.stringify(body),
        headers,
        method: "GET",
    });
    expect(result).toEqual({
        someThing: "Something else.",
    });
});

test("`wrappedFetch` with broken json", async () => {
    const mockCall = jest.fn();
    global.fetch = mockCall;
    const error = new Error();
    mockCall.mockReturnValue({
        json: () => { throw error; },
        ok: true,
    });
    const mockErrorHandler = jest.fn();

    controller.configure({
        errorHandler: mockErrorHandler,
    });

    await expect(controller.wrappedFetch(route, urlParameters, body, query)).rejects.toEqual(error);
    expect(mockErrorHandler).toHaveBeenCalledWith(error);
});

test("`wrappedFetch` with a non-2xx status code", async () => {
    const mockCall = jest.fn();
    global.fetch = mockCall;
    const error = new ApiError(400, { message: "Something went wrong." });
    mockCall.mockReturnValue({
        json: () => ({
            message: "Something went wrong.",
        }),
        ok: false,
        status: 400,
    });
    const mockErrorHandler = jest.fn();

    controller.configure({
        errorHandler: mockErrorHandler,
    });

    await expect(controller.wrappedFetch(route, urlParameters, body, query)).rejects.toEqual(error);
    expect(mockErrorHandler).toHaveBeenCalledWith(error);
});
