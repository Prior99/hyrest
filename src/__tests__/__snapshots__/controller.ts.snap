// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`wrappedFetch\` with a \`.dump()\` route 1`] = `
User {
  "email": "test@example.com",
  "password": "asdfsadf",
}
`;

exports[`\`wrappedFetch\` with a \`.dump()\` route and an array 1`] = `
Array [
  User {
    "email": "test1@example.com",
  },
  User {
    "email": "test2@example.com",
  },
  User {
    "email": "test3@example.com",
  },
]
`;
